File: web-ui/src/api/accounts.ts
import { http } from '@/lib/http'

export interface AccountItem {
  name: string
  path: string
}

export interface AccountDetail extends AccountItem {
  content: string
}

export async function listAccounts(): Promise<AccountItem[]> {
  return await http('/api/accounts')
}

export async function getAccount(name: string): Promise<AccountDetail> {
  return await http(`/api/accounts/${encodeURIComponent(name)}`)
}

export async function createAccount(payload: { name: string; content: string }): Promise<AccountDetail> {
  return await http('/api/accounts', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(payload),
  })
}

export async function updateAccount(name: string, content: string): Promise<AccountDetail> {
  return await http(`/api/accounts/${encodeURIComponent(name)}`, {
    method: 'PUT',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ content }),
  })
}

export async function deleteAccount(name: string): Promise<{ message: string }> {
  return await http(`/api/accounts/${encodeURIComponent(name)}`, { method: 'DELETE' })
}

export async function authorizeAccount(name: string, timeout = 300): Promise<{ message: string; path: string; cookies: number }> {
  return await http(`/api/accounts/${encodeURIComponent(name)}/authorize`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ timeout }),
  })
}

export interface QrSession {
  success: boolean
  session_id: string
  qr_code_url: string
  expires_in: number
  message?: string
}

export interface QrSessionStatus {
  status: string
  session_id: string
  account_name?: string
  message?: string
  verification_url?: string
  path?: string
  cookies?: number
}

export async function createQrSession(name: string): Promise<QrSession> {
  return await http('/api/accounts/qr/session', {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify({ name }),
  })
}

export async function getQrSessionStatus(sessionId: string): Promise<QrSessionStatus> {
  return await http(`/api/accounts/qr/${encodeURIComponent(sessionId)}/status`)
}

File: src/api/routes/accounts.py
"""
闲鱼账号管理路由
"""
import json
import os
import re
import aiofiles
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import List
from src.infrastructure.config.env_manager import env_manager
from src.services.browser_auth import (
    AccountAuthorizationError,
    browser_authorizer,
)
from src.services.qr_login import qr_login_manager


router = APIRouter(prefix="/api/accounts", tags=["accounts"])

ACCOUNT_NAME_RE = re.compile(r"^[a-zA-Z0-9_-]{1,50}$")


class AccountCreate(BaseModel):
    name: str
    content: str


class AccountUpdate(BaseModel):
    content: str


def _strip_quotes(value: str) -> str:
    if not value:
        return value
    if value.startswith(("\"", "'")) and value.endswith(("\"", "'")):
        return value[1:-1]
    return value


def _state_dir() -> str:
    raw = env_manager.get_value("ACCOUNT_STATE_DIR", "state") or "state"
    return _strip_quotes(raw.strip())


def _ensure_state_dir(path: str) -> None:
    os.makedirs(path, exist_ok=True)


def _validate_name(name: str) -> str:
    trimmed = name.strip()
    if not trimmed or not ACCOUNT_NAME_RE.match(trimmed):
        raise HTTPException(status_code=400, detail="账号名称只能包含字母、数字、下划线或短横线。")
    return trimmed


def _account_path(name: str) -> str:
    filename = f"{name}.json"
    return os.path.join(_state_dir(), filename)


def _validate_json(content: str) -> None:
    try:
        json.loads(content)
    except json.JSONDecodeError:
        raise HTTPException(status_code=400, detail="提供的内容不是有效的JSON格式。")


@router.get("", response_model=List[dict])
async def list_accounts():
    state_dir = _state_dir()
    if not os.path.isdir(state_dir):
        return []
    files = [f for f in os.listdir(state_dir) if f.endswith(".json")]
    accounts = []
    for filename in sorted(files):
        name = filename[:-5]
        accounts.append({
            "name": name,
            "path": os.path.join(state_dir, filename),
        })
    return accounts


@router.get("/{name}", response_model=dict)
async def get_account(name: str):
    account_name = _validate_name(name)
    path = _account_path(account_name)
    if not os.path.exists(path):
        raise HTTPException(status_code=404, detail="账号不存在")
    async with aiofiles.open(path, "r", encoding="utf-8") as f:
        content = await f.read()
    return {"name": account_name, "path": path, "content": content}


@router.post("", response_model=dict)
async def create_account(data: AccountCreate):
    account_name = _validate_name(data.name)
    _validate_json(data.content)
    state_dir = _state_dir()
    _ensure_state_dir(state_dir)
    path = _account_path(account_name)
    if os.path.exists(path):
        raise HTTPException(status_code=409, detail="账号已存在")
    async with aiofiles.open(path, "w", encoding="utf-8") as f:
        await f.write(data.content)
    return {"message": "账号已添加", "name": account_name, "path": path}


@router.put("/{name}", response_model=dict)
async def update_account(name: str, data: AccountUpdate):
    account_name = _validate_name(name)
    _validate_json(data.content)
    state_dir = _state_dir()
    _ensure_state_dir(state_dir)
    path = _account_path(account_name)
    if not os.path.exists(path):
        raise HTTPException(status_code=404, detail="账号不存在")
    async with aiofiles.open(path, "w", encoding="utf-8") as f:
        await f.write(data.content)
    return {"message": "账号已更新", "name": account_name, "path": path}


@router.delete("/{name}", response_model=dict)
async def delete_account(name: str):
    account_name = _validate_name(name)
    path = _account_path(account_name)
    if not os.path.exists(path):
        raise HTTPException(status_code=404, detail="账号不存在")
    os.remove(path)
    return {"message": "账号已删除"}


class AccountAuthorize(BaseModel):
    """浏览器授权请求体"""
    timeout: int = 300


class QrSessionCreate(BaseModel):
    """扫码登录会话创建请求体"""
    name: str


@router.post("/qr/session", response_model=dict)
async def create_qr_session(data: QrSessionCreate):
    """生成闲鱼登录二维码，后台监控扫码状态，登录成功后自动保存账号登录态。"""
    account_name = _validate_name(data.name)
    state_dir = _state_dir()
    _ensure_state_dir(state_dir)
    path = _account_path(account_name)
    result = await qr_login_manager.create_session(account_name, path)
    if not result.get("success"):
        raise HTTPException(status_code=400, detail=result.get("message", "生成二维码失败"))
    return result


@router.get("/qr/{session_id}/status", response_model=dict)
async def get_qr_session_status(session_id: str):
    """查询扫码登录会话状态。"""
    return qr_login_manager.get_status(session_id)


@router.post("/{name}/authorize", response_model=dict)
async def authorize_account(name: str, data: AccountAuthorize):
    """打开真实浏览器窗口，引导用户登录闲鱼，登录成功后导出登录态到账号文件。"""
    account_name = _validate_name(name)
    state_dir = _state_dir()
    _ensure_state_dir(state_dir)
    path = _account_path(account_name)
    try:
        result = await browser_authorizer.authorize(path, timeout_sec=data.timeout)
    except AccountAuthorizationError as e:
        raise HTTPException(status_code=400, detail=str(e))
    return {
        "message": "账号授权成功，登录态已保存",
        "name": account_name,
        "path": result["path"],
        "cookies": result["cookies"],
    }
